cpg-library/src/main/java/de/fraunhofer/aisec/cpg/frontends/typescript/TypeScriptLanguageFrontend.kt

```diff
@@ -0,0 +1,190 @@
+import org.checkerframework.checker.nullness.qual.NonNull
+
+/**
+ * This language frontend adds experimental support for TypeScript. It is definitely not feature
+ * complete, but can be used to parse simple typescript snippets through the official typescript
+ * parser written in TypeScript / nodejs. It includes a simple nodejs script that invokes this
+ * parser in `src/main/nodejs`. It basically dumps the AST in a JSON structure on stdout and this
+ * input is parsed by this frontend.
+ *
+ * Because TypeScript is a strict super-set of JavaScript, this frontend can also be used to parse
+ * JavaScript. However, this is not properly tested. Furthermore, the official TypeScript parser
+ * also has built-in support for React dialects TSX and JSX.
+ */
+@ExperimentalTypeScript
+class TypeScriptLanguageFrontend(
+    config: @NonNull TranslationConfiguration,
+    scopeManager: ScopeManager?
+) : LanguageFrontend(config, scopeManager, ".") {
+
+    val declarationHandler = DeclarationHandler(this)
+    val statementHandler = StatementHandler(this)
+    val expressionHandler = ExpressionHandler(this)
+    val typeHandler = TypeHandler(this)
+
+    val mapper = jacksonObjectMapper()
+
+    companion object {
+        @kotlin.jvm.JvmField var TYPESCRIPT_EXTENSIONS: List<String> = listOf(".ts", ".tsx")
+
+        @kotlin.jvm.JvmField var JAVASCRIPT_EXTENSIONS: List<String> = listOf(".js", ".jsx")
+
+        private val parserFile: File = createTempFile("parser", ".js")
+
+        init {
+            val link = this::class.java.getResourceAsStream("/nodejs/parser.js")
+            link?.use {
+                log.info(
+                    "Extracting parser.js out of resources to {}",
+                    parserFile.absoluteFile.toPath()
+                )
+                Files.copy(
+                    it,
+                    parserFile.absoluteFile.toPath(),
+                    StandardCopyOption.REPLACE_EXISTING
+                )
+            }
+        }
+    }
+
+    override fun parse(file: File): TranslationUnitDeclaration {
+        if (!parserFile.exists()) {
+            throw TranslationException("parser.js not found @ ${parserFile.absolutePath}")
+        }
+
+        val p =
+            Runtime.getRuntime().exec(arrayOf("node", parserFile.absolutePath, file.absolutePath))
+
+        val node = mapper.readValue(p.inputStream, TypeScriptNode::class.java)
+
+        return this.declarationHandler.handle(node) as TranslationUnitDeclaration
+    }
+
+    override fun <T : Any?> getCodeFromRawNode(astNode: T): String? {
+        return if (astNode is TypeScriptNode) {
+            return astNode.code
+        } else {
+            null
+        }
+    }
+
+    override fun <T : Any?> getLocationFromRawNode(astNode: T): PhysicalLocation? {
+        return if (astNode is TypeScriptNode) {
+            // TODO: LSP region
+            val location = PhysicalLocation(File(astNode.location.file).toURI(), Region())
```

Comment:
Is this still using the default region?

Reply:
Yes, this is a bit more complicated because of the different code region formats. I would do this in a separate PR.